File: stacks_and_queues/stack/stack.py
class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.next = None


class Stack:

    def __init__(self):
        self.top = None
        self._size = 0

    # insert new item
    def push(self, item):
        node = Node(item)
        node.next = self.top
        self.top = node
        self._size += 1

    # return item
    def pop(self):
        node = self.top
        self.top = self.top.next  # set new top with the right below
        if self._size > 0:
            self._size -= 1
            return node.data
        else:
            raise IndexError("Stack has 0 items")

    # return item, without popping it
    def peek(self):
        if self._size > 0:
            return self.top.data
        else:
            raise IndexError("Stack has 0 itemns")

    def __len__(self):
        return self._size

    def __repr__(self):
        txt, counter = "", 0
        pointer = self.top

        while counter < self._size:
            txt += f" {pointer.data} "
            pointer = pointer.next
            counter += 1

        return txt

    def __eq__(self, other_stack):
        if self._size != other_stack._size:
            return False

        counter = 0
        pointer_a, pointer_b = self.top, other_stack.top

        while counter < self._size and pointer_b.data == pointer_a.data:
            counter += 1
            pointer_a, pointer_b = pointer_a.next, pointer_b.next

        return counter == self._size

File: stacks_and_queues/stack/stack.cpp
#include <iostream>

using namespace std; 


class Node{
    public:
        int data; 
        Node* next; 
        
        Node(int data): data(data), next(nullptr) {}
};


class Stack{
    private:
        Node* top; 
        int _size = 0;

    public:
        void push(int item){
            Node* node = new Node(item);
            node->next = top;
            top = node; 
            _size ++; 
        }

        int pop(){
            Node* node = top; 
            top = top->next;

            if(_size > 0){
                _size --; 
                return node->data; 
            }
            else{
                cerr << "stack has 0 items" << endl;
                return 0; 
            }
        }

        int peek(){
            if(_size > 0){
                return top->data; 
            }
            else{
                cerr << "stack has 0 items" << endl; 
                return 0; 
            }
        }

        int size(){
            return _size; 
        } 

        string repr() const{
            string txt; 
            int counter = 0; 
            Node* pointer = top; 

            while(_size > counter){
                txt += to_string(pointer->data) + " "; 
                pointer = pointer->next; 
                counter++; 
            }
            return txt;
        }

        bool operator==(const Stack& other_stack) const{
            if(_size != other_stack._size){
                return false; 
            }
            
            Node* pointer_a = top;
            Node* pointer_b = other_stack.top; 
            int counter = 0; 

            while(counter < _size && pointer_a->data == pointer_b->data){
                counter++; 
                pointer_a = pointer_a->next; 
                pointer_b = pointer_b->next; 
            }

            return (counter == _size);
        }
};
